replica/replica.go

```diff
@@ -149,34 +157,110 @@ func (replica *Replica) Start() {
+		// in the future, then ignore them. The total window of time is 20
+		// seconds, approximately the latency expected for globally distributed
+		// message passing.
+		now := block.Timestamp(time.Now().Unix())
+		timestamp := m.Message.(*process.Resync).Timestamp()
+		delta := now - timestamp
+		if delta < 0 {
+			delta = -delta
+		}
+		if delta > 10 {
+			replica.options.Logger.Debugf("ignore message: resync timestamp=%v compared to now=%v", timestamp, now)
+			return
+		}
+	}
+
 	// Check that the Message sender is from our Shard (this can be a moderately
 	// expensive operation, so we cache the result until a new `block.Base` is
 	// detected)
 	replica.cache.fillBaseBlock(replica.blockStorage.LatestBaseBlock(replica.shard))
 	if !replica.cache.signatoryInBaseBlock(m.Message.Signatory()) {
 		return
 	}
-
 	// Verify that the Message is actually signed by the claimed `id.Signatory`
 	if err := process.Verify(m.Message); err != nil {
 		replica.options.Logger.Warnf("bad message: unverified: %v", err)
 		return
 	}
 
-	// Handle the underlying `process.Message` and immediately save the
-	// `process.Process` afterwards to protect against unexpected crashes
-	replica.p.HandleMessage(m.Message)
-	replica.p.Save()
+	// Make sure that the Process state gets saved.
+	defer replica.p.Save()
+
+	// Messages from the current height can be handled immediately.
+	if m.Message.Height() == replica.p.CurrentHeight() {
+		replica.p.HandleMessage(m.Message)
+	}
+
+	// Messages from the future must be put into the height-ordered message
+	// queue.
+	if m.Message.Height() > replica.p.CurrentHeight() {
+		if m.Message.Type() != process.ProposeMessageType {
+			// We only want to queue non-Propose messages, because Propose messages
+			// can have a LatestCommit message to fast-forward the process.
+			replica.messageQueue.Push(m.Message)
+		} else {
+			// If the Propose is not at the next height, then we need to make
+			// sure that no base blocks have been missed. Otherwise, reject the
+			// Propose, and wait until the appropriate one has been seen.
+			baseBlockHash := replica.blockStorage.LatestBaseBlock(m.Shard).Hash()
+			blockHash := m.Message.BlockHash()
+			numMissingBaseBlocks := replica.rebaser.blockIterator.BaseBlocksInRange(baseBlockHash, blockHash)
+			if numMissingBaseBlocks == 0 {
+				// If we have missed a base block, we drop the Propose. The
+				// Propose that justifies the next base block will eventually be
+				// seen by this Replica and we can begin accepting Proposes from
+				// the new base.
+
+				// In this condition, we haven't missed any base blocks, so we
+				// can proceed as usual.
+				replica.p.HandleMessage(m.Message)
+			}
+		}
+	}
+
+	queued := replica.messageQueue.PopUntil(replica.p.CurrentHeight())
+	for queued != nil && len(queued) > 0 {
+		for _, message := range queued {
+			// Handle the underlying `process.Message` and immediately save the
```

Comment:
Just to be clear, it doesn't seem we're saving the process after each message. We're only saving it once all messages are handled.

Reply:
That's correct. This is to be slightly more efficient. Process saving is actually not about "saving after each message". It's about making sure that if we crash part way through handling a message, we haven't partially saved the Process state (can be caused by "saving as we go").

Reply:
Yep, that's what I figured. Could be worth updating the comment to reflect that since it doesn't exactly reflect what's happening.

Reply:
Done.